File: Xamarin_Forms/Xamarin_Forms/GertecXamarinForms/GertecXamarinForms.Android/SAT/ServiceSat/OperacaoSat.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using GertecXamarinForms.Droid.Sat;

namespace GertecXamarinForms.Droid.SAT.ServiceSat
{
    public class OperacaoSat
    {

        public static RetornoSat invocarOperacaoSat(string operacao, string respostaSat)
        {
            return new RetornoSat(operacao, respostaSat);
        }

        // Função que retorna uma String formatada para ser posta no dialogo Sat, com todas as informações do retorno da operação realizada
        // Em caso de erro retorna o erro
        // *Existem operações que vão ter como resultado informações semelhantes
        public static string formataRetornoSat(RetornoSat retornoSat)
        {
            string retornoFormatado = "";
            // Verifica se existe um erro no retorno
            if (!(retornoSat.ErroSat == ""))
            {
                // Retorna o erro caso exista
                return "Mensagem: " + retornoSat.ErroSat;
            }
            else
            {
                //* Estas informações são padrões a todos os retornos(sempre vão aparecer no Dialogo Sat)
                //* Para mais informações consulte o arquivo retornosSat.txt. É possivel visualizar nele a posição e informação de cada Retorno do Sat.
                if (!(retornoSat.NumeroCCCC == ""))
                {
                    // Se tiver Código CCCC adiciona ele a string formatada
                    retornoFormatado = retornoSat.NumeroEEEE + "|" + retornoSat.NumeroCCCC + "-";
                }
                else
                {
                    // Caso não tenha o código CCCC adicionada somente o EEEE
                    retornoFormatado = retornoSat.NumeroEEEE + "-";
                }

                retornoFormatado += retornoSat.Mensagem;

                retornoFormatado += "\n"; // Pula linha

                // Verifica se adiciona o Código e Mensagem Sefaz na mensagem, caso não estejam vazios
                if (!(retornoSat.NumeroCod == "" || retornoSat.MensagemSefaz == ""))
                {
                    retornoFormatado += "Cod/Mens Sefaz: \n" + retornoSat.NumeroCod + "-" + retornoSat.MensagemSefaz;
                }
                else if (!(retornoSat.NumeroCod == ""))
                {
                    retornoFormatado += "Cod/Mens Sefaz: \n" + retornoSat.NumeroCod + "-";
                }
                else if (!(retornoSat.MensagemSefaz == ""))
                {
                    retornoFormatado += "Cod/Mens Sefaz: \n" + "-" + retornoSat.MensagemSefaz;
                }

                retornoFormatado += "\n"; // Pula linha

                //* Agora só são inseridas as informações que não são padrões a todos retornos
                //* São atribuidas as informações especificas do retorno da operação
                if (retornoSat.Operacao == "AtivarSAT")
                {
                    if (!(retornoSat.CodigoCSR == ""))
                    {
                        retornoFormatado += "CSR: " + retornoSat.CodigoCSR;
                    }
                }
                else if (retornoSat.Operacao == "ExtrairLog")
                {
                    //! Cuidado com está parte, ela pode  exigir muito processamento se estiver em modo Debug
                    //! Recomenda-se que utilize somente em modo release e não colocar em um Dialogo, pois o arquivo retornado é grande
                    // retornoFormatado += "Arquivo de log em base64: " + retornoSat.getLogBase64;
                }
                else if (retornoSat.Operacao == "ConsultarStatusOperacional")
                {
                    retornoFormatado += "------- Conteúdo Retorno -------" +
                            "\n" +
                            "Número de Série do SAT: " +
                            retornoSat.NumeroSerieSat +
                            "Tipo de Lan: " +
                            retornoSat.TipoLan +
                            "\n" +
                            "IP SAT: " +
                            retornoSat.IpSat +
                            "\n" +
                            "MAC SAT: " +
                            retornoSat.MacSat +
                            "\n" +
                            "Máscara: " +
                            retornoSat.Mascara +
                            "\n" +
                            "Gateway: " +
                            retornoSat.Gateway +
                            "\n" +
                            "DNS 1: " +
                            retornoSat.Dns1 +
                            "\n" +
                            "DNS 2: " +
                            retornoSat.Dns2 +
                            "\n" +
                            "Status da Rede: " +
                            retornoSat.StatusRede +
                            "\n" +
                            "Nível da Bateria: " +
                            retornoSat.NivelBateria +
                            "\n" +
                            "Memória de Trabalho Total: " +
                            retornoSat.MemoriaDeTrabalhoTotal +
                            "\n" +
                            "Memória de Trabalho Usada: " +
                            retornoSat.MemoriaDeTrabalhoUsada +
                            "\n" +
                            "Data/Hora: " +
                            retornoSat.DataHora +
                            "\n" +
                            "Versão: " +
                            retornoSat.Versao +
                            "\n" +
                            "Versão de Leiaute: " +
                            retornoSat.VersaoLeiaute +
                            "\n" +
                            "Último CFe-Sat Emitido: " +
                            retornoSat.UltimoCfeEmitido +
                            "\n" +
                            "Primeiro CFe-Sat Em Memória: " +
                            retornoSat.PrimeiroCfeMemoria +
                            "\n" +
                            "Último CFe-Sat Em Memória: " +
                            retornoSat.UltimoCfeMemoria +
                            "\n" +
                            "Última Transmissão de CFe-SAT para SEFAZ: " +
                            retornoSat.UltimaTransmissaoSefazDataHora +
                            "\n" +
                            "Última Comunicacao com a SEFAZ:" +
                            retornoSat.UltimaComunicacaoSefazData +
                            "\n" +
                            "Data de emissão do certificado: " +
                            retornoSat.DataEmissaoCertificado +
                            "\n" +
                            "Data de vencimento do certificado: " +
                            retornoSat.DataVencimentoCertificado +
                            "\n" +
                            "Estado de Operação do SAT: " +
                            retornoSat.EstadoDeOperacao;
                }
                else if (retornoSat.Operacao == "AssociarSAT")
                {
                    //* Associar SAT somente tem como dado especifico o campo CCCC(fica ao seu criterio adicionar-lo ou não)
                }
                else if (retornoSat.Operacao == "EnviarTesteFim")
                {
                    retornoFormatado += "TimeStamp: " +
                            retornoSat.TimeStamp +
                            "\nNum Doc Fiscal: " +
                            retornoSat.NumDocFiscal +
                            "\nChave de Consulta: " +
                            retornoSat.ChaveConsulta +
                            "\nArquivo CFE Base 64: " +
                            ConverterBase64EmXml(retornoSat.ArquivoCFeBase64);
                }
                else if ((retornoSat.Operacao == "EnviarTesteVendas") || (retornoSat.Operacao == "CancelarUltimaVenda"))
                {
                    retornoFormatado += "TimeStamp: " +
                            retornoSat.TimeStamp +
                            "\nChave de Consulta: " +
                            retornoSat.ChaveConsulta +
                            "\nValor CFE: " +
                            retornoSat.ValorTotalCFe +
                            "\nValor CPF CNPJ: " +
                            retornoSat.CPFCNPJValue +
                            "\nArquivo CFE Base 64: " +
                            ConverterBase64EmXml(retornoSat.ArquivoCFeBase64) +
                            "\nAssinatura QRCODE: " +
                            retornoSat.AssinaturaQRCODE;
                }
            }
            return retornoFormatado;
        }

        public static string ConverterBase64EmXml(string base64Sat)
        {
            byte[] data = Convert.FromBase64String(base64Sat);
            return System.Text.ASCIIEncoding.ASCII.GetString(data);
        }

    }
}

File: Android_Nativo_Java/Android/app/src/main/java/com/gertec/exemplosgertec/ExemploSAT/ServiceSAT/OperacaoSat.java
package com.gertec.exemplosgertec.ExemploSAT.ServiceSAT;

import android.util.Base64;

import java.io.UnsupportedEncodingException;
import java.util.Map;

public class OperacaoSat {

    public static RetornoSat invocarOperacaoSat(String operacao, String respostaSat) {
        return new RetornoSat(operacao, respostaSat);
    }

    // Função que retorna uma String formatada para ser posta no dialogo Sat, com todas as informações do retorno da operação realizada
    // Em caso de erro retorna o erro
    // *Existem operações que vão ter como resultado informações semelhantes
    public static String formataRetornoSat(RetornoSat retornoSat) {
        String retornoFormatado = "";
        // Verifica se existe um erro no retorno
        if (retornoSat.getErroSat().equals("") == false) {
            // Retorna o erro caso exista
            return "Mensagem: " + retornoSat.getErroSat();
        } else {
            //* Estas informações são padrões a todos os retornos(sempre vão aparecer no Dialogo Sat)
            //* Para mais informações consulte o arquivo retornosSat.txt. É possivel visualizar nele a posição e informação de cada Retorno do Sat.
            if (retornoSat.getNumeroCCCC().equals("") == false) {
                // Se tiver Código CCCC adiciona ele a string formatada
                retornoFormatado = retornoSat.getNumeroEEEE() + "|" + retornoSat.getNumeroCCCC() + "-";
            } else {
                // Caso não tenha o código CCCC adicionada somente o EEEE
                retornoFormatado = retornoSat.getNumeroEEEE() + "-";
            }

            retornoFormatado += retornoSat.getMensagem();

            retornoFormatado += "\n"; // Pula linha

            // Verifica se adiciona o Código e Mensagem Sefaz na mensagem, caso não estejam vazios
            if (retornoSat.getNumeroCod().equals("") == false && retornoSat.getMensagemSefaz().equals("") == false) {
                retornoFormatado += "Cod/Mens Sefaz: \n" + retornoSat.getNumeroCod() + "-" + retornoSat.getMensagemSefaz();
            } else if (retornoSat.getNumeroCod().equals("") == false) {
                retornoFormatado += "Cod/Mens Sefaz: \n" + retornoSat.getNumeroCod() + "-";
            } else if (retornoSat.getMensagemSefaz().equals("") == false) {
                retornoFormatado += "Cod/Mens Sefaz: \n" + "-" + retornoSat.getMensagemSefaz();
            }

            retornoFormatado += "\n"; // Pula linha

            //* Agora só são inseridas as informações que não são padrões a todos retornos
            //* São atribuidas as informações especificas do retorno da operação
            if (retornoSat.getOperacao().equals("AtivarSAT")) {
                if (retornoSat.getCodigoCSR().equals("") == false) {
                    retornoFormatado += "CSR: " + retornoSat.getCodigoCSR();
                }
            } else if (retornoSat.getOperacao().equals("ExtrairLog")) {
                //! Cuidado com está parte, ela pode  exigir muito processamento se estiver em modo Debug
                //! Recomenda-se que utilize somente em modo release e não colocar em um Dialogo, pois o arquivo retornado é grande
                // retornoFormatado += "Arquivo de log em base64: " + retornoSat.getLogBase64;
            } else if (retornoSat.getOperacao().equals("ConsultarStatusOperacional")) {
                retornoFormatado += "------- Conteúdo Retorno -------" +
                        "\n" +
                        "Número de Série do SAT: " +
                        retornoSat.getNumeroSerieSat() +
                        "Tipo de Lan: " +
                        retornoSat.getTipoLan() +
                        "\n" +
                        "IP SAT: " +
                        retornoSat.getIpSat() +
                        "\n" +
                        "MAC SAT: " +
                        retornoSat.getMacSat() +
                        "\n" +
                        "Máscara: " +
                        retornoSat.getMascara() +
                        "\n" +
                        "Gateway: " +
                        retornoSat.getGateway() +
                        "\n" +
                        "DNS 1: " +
                        retornoSat.getDns1() +
                        "\n" +
                        "DNS 2: " +
                        retornoSat.getDns2() +
                        "\n" +
                        "Status da Rede: " +
                        retornoSat.getStatusRede() +
                        "\n" +
                        "Nível da Bateria: " +
                        retornoSat.getNivelBateria() +
                        "\n" +
                        "Memória de Trabalho Total: " +
                        retornoSat.getMemoriaDeTrabalhoTotal() +
                        "\n" +
                        "Memória de Trabalho Usada: " +
                        retornoSat.getMemoriaDeTrabalhoUsada() +
                        "\n" +
                        "Data/Hora: " +
                        retornoSat.getDataHora() +
                        "\n" +
                        "Versão: " +
                        retornoSat.getVersao() +
                        "\n" +
                        "Versão de Leiaute: " +
                        retornoSat.getVersaoLeiaute() +
                        "\n" +
                        "Último CFe-Sat Emitido: " +
                        retornoSat.getUltimoCfeEmitido() +
                        "\n" +
                        "Primeiro CFe-Sat Em Memória: " +
                        retornoSat.getPrimeiroCfeMemoria() +
                        "\n" +
                        "Último CFe-Sat Em Memória: " +
                        retornoSat.getUltimoCfeMemoria() +
                        "\n" +
                        "Última Transmissão de CFe-SAT para SEFAZ: " +
                        retornoSat.getUltimaTransmissaoSefazDataHora() +
                        "\n" +
                        "Última Comunicacao com a SEFAZ:" +
                        retornoSat.getUltimaComunicacaoSefazData() +
                        "\n" +
                        "Estado de Operação do SAT: " +
                        retornoSat.getEstadoDeOperacao();
            } else if (retornoSat.getOperacao().equals("AssociarSAT")) {
                //* Associar SAT somente tem como dado especifico o campo CCCC(fica ao seu criterio adicionar-lo ou não)
            } else if (retornoSat.getOperacao().equals("EnviarTesteFim")) {
                retornoFormatado += "TimeStamp: " +
                        retornoSat.getTimeStamp() +
                        "\nNum Doc Fiscal: " +
                        retornoSat.getNumDocFiscal() +
                        "\nChave de Consulta: " +
                        retornoSat.getChaveConsulta() +
                        "\nArquivo CFE Base 64: " +
                        converterBase64EmXml(retornoSat.getArquivoCFeBase64());
            } else if (retornoSat.getOperacao().equals("EnviarTesteVendas") || retornoSat.getOperacao().equals("CancelarUltimaVenda")) {
                retornoFormatado += "TimeStamp: " +
                        retornoSat.getTimeStamp() +
                        "\nChave de Consulta: " +
                        retornoSat.getChaveConsulta() +
                        "\nValor CFE: " +
                        retornoSat.getValorTotalCFe() +
                        "\nValor CPF CNPJ: " +
                        retornoSat.getCPFCNPJValue() +
                        "\nArquivo CFE Base 64: " +
                        converterBase64EmXml(retornoSat.getArquivoCFeBase64()) +
                        "\nAssinatura QRCODE: " +
                        retornoSat.getAssinaturaQRCODE();
            }
        }
        return retornoFormatado;
    }

    public static String converterBase64EmXml(String base64Sat) {
        String retornoSemBase64 = "";
        try {
            byte[] data = Base64.decode(base64Sat, Base64.DEFAULT);
            retornoSemBase64 = new String(data, "UTF-8");
            return retornoSemBase64;
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return retornoSemBase64;
    }
}
